.gitlab-ci.yml

```diff
@@ -1,34 +1,43 @@
-image: docker:19.03.12stages:
+image: docker:19.03.12
+
+stages:
   - testing
   - release
 
 variables:
+  GITLABCI_VERSION:    "2.1"
   GITLABCI_IMAGE_TEST: $CI_REGISTRY_IMAGE:$CI_COMMIT_SHA
-  GITLABCI_IMAGE:      $CI_REGISTRY_IMAGE:2.1
+  GITLABCI_IMAGE:      $CI_REGISTRY_IMAGE:$GITLABCI_VERSION
 
 check-pr:
-  stage: testing 
+  stage: testing
+  services:
+    - docker:19.03.12-dind
   except:
     - master
-  #only:
-  #  changes:
-  #    - docker-gitlabci/*
+  only:
+    changes:
+      - docker-gitlabci/*
   script:
-    - ./docker-gitlabci/build.sh
+    - cd docker-gitlabci
+    - ./build.sh $GITLABCI_VERSION
 
 build-gitlabci:
+  environment: maintainer-approved
```

Comment:
What does this do?

Reply:
The idea is that you add your special token for dockerhub in gitlab, and only runs triggered by maintainers in gitlab can trigger runs for that environment. In this way an external contributor cannot steal your tokens...

Because when that works, we can than also have that you can trigger the uploading either by merging after the zip is there or by using the "run pipeline" where you add a variable for where the zip can be downloaded from.

TL;DR not needed for this one, but adds some form of obfuscation in such a way that someone cannot push a new gitlabci which isn't tested yet.

Reply:
I thought we could do this with a protected environment variable in GitLab? That we only expose it on master.
I think setting an `environment` key itself doesn't do much? Or do you have an example of what it does?

Reply:
You need both,

https://docs.gitlab.com/ee/ci/environments/protected_environments.html

Currently we do not have an environment yet, (I cannot select it) so we need to add an environment in the tasks, (although I think it can be global instead of what I do now) which will than only expose the variables of username/password when someone with maintainer role (or higher) in gitlab runs this (current list: https://gitlab.com/groups/DOMjudge/-/group_members).

As we discussed to actualy make this a seperate pipeline I can shift this a bit, if you change my access to developer we can even set an environment for gitlabci deployment (allowing me and Mart to also do this) and set another environment for DOMjudge release deployment (allowing this only to maintainers -> All the owners)

TL;DR for the gitlab interface we atleast need 1 environment (there is not a default) and environment !== branch, and you're right, I also need to setup additional things in GitLab

Reply:
I was thinking to use this:
<img width="758" alt="image" src="https://user-images.githubusercontent.com/550145/106441824-d22ec400-647a-11eb-9c16-9b9ebdaf6737.png">
i.e. the 'Protect variables' thing and then make only `master` a protected branch. Then we do not need the environment I think?

Reply:
I think you're right, I'll try to build something in mvasseur for gitlabci tonight.